src/sumcheck/constraints.rs

```diff
@@ -0,0 +1,695 @@
+            bindings = new_bindings;
+        }
+
+        // Output the linear constraint that the final claims match the binding of the inputs:
+        //
+        //      SUM_{i} (eq2[i + npub] * sym_private_inputs[i])
+        //      - sym_layer_pad.vl
+        //      - gamma * sym_layer_pad.vr
+        //    =
+        //      - SUM_{i} (eq2[i] * public_inputs[i])
+        //      + claims[0]
+        //      + gamma * claims[1]
+        //
+        // where sym_layer_pad is the padding on the input layer
+        let gamma = transcript.generate_challenge(1)?[0];
+        let eq2 = bindeq(&bindings[0]).elementwise_sum(&bindeq(&bindings[1]).scale(gamma));
+        let input_layer_index = circuit.num_layers();
+
+        // One linear constraint LHS term for each private input
+        for (private_input_index, private_input_witness) in
+            witness_layout.private_input_witness_indices().enumerate()
+        {
+            constraints
+                .linear_constraint_lhs
+                .push(LinearConstraintLhsTerm {
+                    constraint_number: input_layer_index,
+                    witness_index: private_input_witness,
+                    constant_factor: eq2[private_input_index + circuit.num_public_inputs()],
+                });
+        }
+
+        let (vl_witness, vr_witness, _) =
+            witness_layout.wire_witness_indices(input_layer_index - 1);
+
+        // Linear constraint LHS term for sym_layer_pad.vl
+        constraints
+            .linear_constraint_lhs
+            .push(LinearConstraintLhsTerm {
+                constraint_number: input_layer_index,
+                witness_index: vl_witness,
+                constant_factor: -FE::ONE,
+            });
+
+        // Linear constraint LHS term for sym_layer_pad.vr
+        constraints
+            .linear_constraint_lhs
+            .push(LinearConstraintLhsTerm {
+                constraint_number: input_layer_index,
+                witness_index: vr_witness,
+                constant_factor: -gamma,
+            });
+
+        // Linear constraint RHS
+        constraints.linear_constraint_rhs.push(
+            claims[0] + gamma * claims[1]
+                - public_inputs
+                    .iter()
+                    .zip(eq2.iter())
+                    .fold(FE::ZERO, |sum, (eq2_i, public_input_i)| {
+                        sum + *eq2_i * public_input_i
+                    }),
+        );
+
+        Ok(constraints)
+    }
+}
+
+/// From Section 6.6:
+/// LET lag_i(x) =
+///                the quadratic polynomial such that
+///                       lag_i(P_k) = 1  if i = k
+///                                    0  otherwise
+///                for 0 <= k < 3
+///
+/// https://datatracker.ietf.org/doc/html/draft-google-cfrg-libzk-01#section-6.6
+fn lagrange_basis_polynomial_i<FE: FieldElement>(i: FE, x: FE) -> FE {
+    // Our nodes are x_0 = 0, x_1 = 1, and x_2 = SUMCHECK_P2 (aka 2) in the
```

Comment:
We can't assume that SUMCHECK_P2 is 2, because we'll have to use x on $GF(2^{128})$ instead. Thus, we'd need additional precomputed inverses for each field, especially $(p2 - 1)^{-1}$.